src/main/java/io/vertx/core/streams/impl/OutboundWriteQueue.java

```diff
@@ -0,0 +1,361 @@
+ * producers and the consumer.</p>
+ *
+ * The queue maintains an internal queue of elements initially empty and can be filled when
+ * <ul>
+ *   <li>producer threads {@link #submit)} to the queue above the {@link #highWaterMark}</li>
+ *   <li>the {@link #consumer} refuses an element</li>
+ * </ul>
+ *
+ * <p>When the internal queue grows above the {@link #highWaterMark}, the queue is considered as {@code unwritable}.
+ * {@link #add}/{@link #submit} methods return {@link #QUEUE_UNWRITABLE_MASK} to signal producers should stop emitting.</p>
+ *
+ * <p>After a drain if the internal queue has shrunk under {@link #lowWaterMark}, the queue is considered as {@code writable}.
+ * {@link #add}/{@link #drain} methods return {@link #QUEUE_WRITABLE_MASK} to signal producers should start emitting. Note
+ * the consumer thread handle this signal and should transmit it to the producers.</p>
+ *
+ * <p>When {@link #QUEUE_WRITABLE_MASK} is signalled, the number of {@link #QUEUE_UNWRITABLE_MASK} signals emitted is encoded
+ * in the flags. This number allows the producer flow controller to correctly account the producer writability:
+ * {@link #QUEUE_WRITABLE_MASK}/{@link #QUEUE_UNWRITABLE_MASK} signals are observed by different threads, therefore a good
+ * practice to compute the queue writability is to increment/decrement an atomic counter for each signal received.</p>
+ */
+public class OutboundWriteQueue<E> {
+
+  /**
+   * Returns the number of times {@link #QUEUE_UNWRITABLE_MASK} signals encoded in {@code value}
+   *
+   * @param value the value
+   * @return then number of unwritable signals
+   */
+  public static int numberOfUnwritableSignals(int value) {
+    return (value & ~0XF) >> 4;
+  }
+
+  /**
+   * When the masked bit is set, the queue became unwritable, this triggers only when the queue transitions
+   * from the <i>writable</i>> state to the <i>unwritable</i>> state.
+   */
+  public static int QUEUE_UNWRITABLE_MASK = 0x01;
+
+  /**
+   * When the masked bit is set, the queue became writable, this triggers only when the queue transitions
+   * from the <i>unwritable</i>> state to the <i>writable</i> state.
+   */
+  public static int QUEUE_WRITABLE_MASK = 0x02;
+
+  /**
+   * When the masked bit is set, the caller has acquired the ownership of the queue and must drain it
+   * to attempt to release the ownership.
+   */
+  public static int DRAIN_REQUIRED_MASK = 0x04;
+
+  /**
+   * The default high-water mark value: {@code 16}
+   */
+  public static int DEFAULT_HIGH_WATER_MARK = 16;
+
+  /**
+   * The default low-water mark value: {@code 8}
+   */
+  public static int DEFAULT_LOW_WATER_MARK = 8;
+
+  // Immutable
+
+  private final Predicate<E> consumer;
+  private final int highWaterMark;
+  private final int lowWaterMark;
+
+  // Concurrent part accessed by any producer thread
+
+  private final Queue<E> queue = PlatformDependent.newMpscQueue();
+  private final AtomicInteger wip = new AtomicInteger();
```

Comment:
I would consider using atomic int field updater: in subsequent iterations it could allow padding or just save allocating a wrapper Atomic, in both cases we can save this.
Another important point: given that we have no control on the amount of concurrent, accumulates writes and the OS pauses due to scarse CPU resources, is often better to avoid integer overflow by using a long here. 
External concurrent requests could be a lot, in pathological cases and addressing numeric overflow is a PITA.

Most HW arch words are 8 bytes anyway so...no need to save bits